src/main/java/io/cryostat/net/reports/RemoteReportGenerator.java

```diff
@@ -0,0 +1,125 @@
+ * either these or other terms.
+ *
+ * This license is subject to the following condition:
+ * The above copyright notice and either this complete permission notice or at
+ * a minimum a reference to the UPL must be included in all copies or
+ * substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package io.cryostat.net.reports;
+
+import java.nio.file.Path;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+
+import javax.inject.Provider;
+
+import io.cryostat.configuration.Variables;
+import io.cryostat.core.log.Logger;
+import io.cryostat.core.sys.Environment;
+import io.cryostat.core.sys.FileSystem;
+import io.cryostat.net.TargetConnectionManager;
+import io.cryostat.net.web.http.HttpMimeType;
+import io.cryostat.util.HttpStatusCodeIdentifier;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import io.vertx.core.Vertx;
+import io.vertx.ext.web.client.WebClient;
+import io.vertx.ext.web.multipart.MultipartForm;
+
+class RemoteReportGenerator extends AbstractReportGeneratorService {
+
+    private final Vertx vertx;
+    private final WebClient http;
+    private final Environment env;
+
+    RemoteReportGenerator(
+            TargetConnectionManager targetConnectionManager,
+            FileSystem fs,
+            Provider<Path> tempFileProvider,
+            Vertx vertx,
+            WebClient http,
+            Environment env,
+            Logger logger) {
+        super(targetConnectionManager, fs, tempFileProvider, logger);
+        this.vertx = vertx;
+        this.http = http;
+        this.env = env;
+    }
+
+    @Override
+    @SuppressFBWarnings("NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE")
+    public CompletableFuture<Path> exec(Path recording, Path destination) {
+        String reportGenerator = env.getEnv(Variables.REPORT_GENERATOR_ENV);
+        logger.info("POSTing {} to {}", recording, reportGenerator);
+        var form =
+                MultipartForm.create()
+                        .binaryFileUpload(
+                                "file",
+                                recording.getFileName().toString(),
+                                recording.toAbsolutePath().toString(),
+                                HttpMimeType.OCTET_STREAM.mime());
+        var f = new CompletableFuture<Path>();
+        this.http
+                .postAbs(String.format("%s/report", reportGenerator))
+                .timeout(TimeUnit.MINUTES.toMillis(1))
```

Comment:
When I tried uploading a ~10MB recording to the cryostat pod archives (noop auth), the upload succeeds. Then when I click the dropdown to generate a report, the backend hangs for a little while before I see this exception:
```
Dec 10, 2021 7:45:48 PM io.cryostat.core.log.Logger info
INFO: POSTing /opt/cryostat.d/recordings.d/file-uploads/aa06f216-5dd5-4e64-a9aa-799994659a80 to http://localhost:10001
Dec 10, 2021 7:46:18 PM io.cryostat.core.log.Logger error
SEVERE: Exception thrown
java.util.concurrent.ExecutionException: io.vertx.core.VertxException: Connection was closed
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:395)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1999)
	at io.cryostat.net.reports.ArchivedRecordingReportCache.get(ArchivedRecordingReportCache.java:82)
	at io.cryostat.net.reports.ReportService.get(ReportService.java:57)
	at io.cryostat.net.web.http.api.v1.ReportGetHandler.handleAuthenticated(ReportGetHandler.java:110)
	at io.cryostat.net.web.http.AbstractAuthenticatedRequestHandler.handle(AbstractAuthenticatedRequestHandler.java:90)
	at io.cryostat.net.web.http.AbstractAuthenticatedRequestHandler.handle(AbstractAuthenticatedRequestHandler.java:65)
	at io.vertx.ext.web.impl.BlockingHandlerDecorator.lambda$handle$0(BlockingHandlerDecorator.java:48)
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:313)
	at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: io.vertx.core.VertxException: Connection was closed

Dec 10, 2021 7:46:18 PM io.cryostat.core.log.Logger error
SEVERE: HTTP 500: java.util.concurrent.ExecutionException: io.vertx.core.VertxException: Connection was closed
io.vertx.ext.web.handler.impl.HttpStatusException: Internal Server Error
Caused by: java.util.concurrent.ExecutionException: java.util.concurrent.ExecutionException: io.vertx.core.VertxException: Connection was closed
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:395)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1999)
	at io.cryostat.net.web.http.api.v1.ReportGetHandler.handleAuthenticated(ReportGetHandler.java:110)
	at io.cryostat.net.web.http.AbstractAuthenticatedRequestHandler.handle(AbstractAuthenticatedRequestHandler.java:90)
	at io.cryostat.net.web.http.AbstractAuthenticatedRequestHandler.handle(AbstractAuthenticatedRequestHandler.java:65)
	at io.vertx.ext.web.impl.BlockingHandlerDecorator.lambda$handle$0(BlockingHandlerDecorator.java:48)
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:313)
	at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.util.concurrent.ExecutionException: io.vertx.core.VertxException: Connection was closed
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:395)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1999)
	at io.cryostat.net.reports.ArchivedRecordingReportCache.get(ArchivedRecordingReportCache.java:82)
	at io.cryostat.net.reports.ReportService.get(ReportService.java:57)
	... 10 more
Caused by: io.vertx.core.VertxException: Connection was closed
Dec 10, 2021 7:46:18 PM io.cryostat.core.log.Logger info
INFO: (10.0.2.100:48070): GET /api/v1/reports/aa06f216-5dd5-4e64-a9aa-799994659a80 500 29582ms
```
If I click View Report, the new tab keeps loading and the backend hangs with the latest logs here:
```
Dec 10, 2021 7:56:02 PM io.cryostat.core.log.Logger info
INFO: (10.0.2.100:48106): POST /api/beta/auth/token 200 9ms
Dec 10, 2021 7:56:02 PM io.cryostat.core.log.Logger info
INFO: POSTing /opt/cryostat.d/recordings.d/file-uploads/aa06f216-5dd5-4e64-a9aa-799994659a80 to http://localhost:10001
```
I don't see this behaviour with a smaller ~6MB recording. Does this happen to you?

Here's the 10MB recording I was using:
[allocation_before_gc_fix.zip](https://github.com/cryostatio/cryostat/files/7695124/allocation_before_gc_fix.zip)

Reply:
Could you check the logs at `podman logs -f reports` to see what's happening in the cryostat-reports container?

Reply:
Either way it sounds like I need to apply a fix so that it doesn't hang at least, but I'm curious what's happening in the report generator container to cause it to fail to respond.

Reply:
The pod looks like it terminated:
```
> podman logs -f reports
exec java -Dquarkus.http.host=0.0.0.0 -Djava.util.logging.manager=org.jboss.logmanager.LogManager -XX:+ExitOnOutOfMemoryError -cp . -jar /deployments/quarkus-run.jar
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/deployments/lib/main/org.jboss.slf4j.slf4j-jboss-logmanager-1.1.0.Final.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/deployments/lib/main/org.slf4j.slf4j-jdk14-1.7.30.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.Slf4jLoggerFactory]
__  ____  __  _____   ___  __ ____  ______ 
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/ 
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \   
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/   
2021-12-10 20:56:39,161 INFO  [io.quarkus] (main) cryostat-reports 1.0.0-SNAPSHOT on JVM (powered by Quarkus 2.4.2.Final) started in 3.367s. Listening on: http://0.0.0.0:10001
2021-12-10 20:56:39,168 INFO  [io.quarkus] (main) Profile prod activated. 
2021-12-10 20:56:39,168 INFO  [io.quarkus] (main) Installed features: [cdi, resteasy-reactive, smallrye-context-propagation, vertx]
exec java -Dquarkus.http.host=0.0.0.0 -Djava.util.logging.manager=org.jboss.logmanager.LogManager -XX:+ExitOnOutOfMemoryError -cp . -jar /deployments/quarkus-run.jar
jalaw@localhost ~/cryostat-operator (main)> 
```

Reply:
What happens if you increase the `--cpus` and `--memory` flag in `smoketest.sh` on the `reports` container? Or even just remove those flags altogether?

Reply:
>  Or even just remove those flags altogether?

If I upload the recording, then click View Report, it hangs.

If I upload the recoridng, then click the dropdown, the report shows up nicely in the dropdown after 5s, and a subsequent click on View Reports show the report in a new tab.

Reply:
That's a little strange. Thanks for checking for me - I'll take a closer look into this soon.

Reply:
Testing it now with your `allocation_before_gc_fix.jfr`. That should fail to upload to archives because the name pattern doesn't match, but it looks like it gets rejected with a `400` and yet the original UUID-named file corresponding to it is still present on the server. That's a bug in the V1 `RecordingsPostHandler`. I have renamed it to be similar to one of my other sample recordings that does have the correct name pattern, so now it is `janlaw-6f4775cdbf-82dvl_10mb_20211006T152006Z.jfr`. I'll see what's going on with this, but it definitely looks like the report generator container is just getting killed and restarted.

Reply:
Reviewing the -web implementation,

> If I upload the recording, then click View Report, it hangs.

This goes through the newer `/api/beta` endpoint for report generation using a JWT token.

> If I upload the recording, then click the dropdown, the report shows up nicely in the dropdown after 5s, and a subsequent click on View Reports show the report in a new tab.

This still goes through the old `/api/v1` endpoint using the `reportUrl` given by `GET /api/v1/recordings`.

Could explain the difference in behaviour. They both end up calling through to the same report generator service which is behind the same archived report cache, which is why in the second case doing the dropdown first causes the subsequent "View Reports" to successfully open (and probably very quickly).

Reply:
@jan-law the hanging problem should be fixed with the latest commit I just added. This applies a timeout to all of the `Future.get()`s involved in report generation, so it should take at most ~30s to get a response, and no more indefinite hang.

Reply:
As for the problem of the sidecar dying when run with `--cpus 1 --memory 512M`, it appears it's getting OOM killed. That seems pretty weird - it's generating the reports the same way that the SubprocessReportGenerator does it, so I can't see why it's requiring so much memory. `run.sh` runs Cryostat itself with `--memory 512M` so it seems very suspicious that the subprocess generator can make do with whatever memory is available to it, but the sidecar version can't do it when it's the only thing running in a container with the same memory limit. I would hazard a guess that there is some kind of Quarkus/Vertx/Netty file upload handling or multipart decoding problem going on here again. In any case, that's a bug report for cryostatio/cryostat-reports.

Reply:
Is the merging of this PR now dependent on this sidecar OOM bug being fixed first?

Reply:
I think so. I haven't gotten around to checking out what's going on with the sidecar container, but there is no point to this PR until the sidecar is useful in practice.

Reply:
> Is the merging of this PR now dependent on this sidecar OOM bug being fixed first?

Resolving this conversation because https://github.com/cryostatio/cryostat-reports/pull/7 is merged